File: cpp/src/order_book.cpp
#include "llab/order_book.hpp"

#include <algorithm>
#include <array>
#include <stdexcept>
#include <vector>

namespace llab::order_book {
namespace {

constexpr std::uint64_t kFnvOffsetBasis = 14695981039346656037ULL;
constexpr std::uint64_t kFnvPrime = 1099511628211ULL;

void hash_byte(std::uint64_t &hash, std::uint8_t byte) {
    hash ^= byte;
    hash *= kFnvPrime;
}

void hash_u64(std::uint64_t &hash, std::uint64_t value) {
    for (unsigned shift = 0; shift < 64; shift += 8) {
        hash_byte(hash, static_cast<std::uint8_t>(value >> shift));
    }
}

std::uint64_t mix_order_id(std::uint64_t value) {
    value += 0x9E3779B97F4A7C15ULL;
    value = (value ^ (value >> 30U)) * 0xBF58476D1CE4E5B9ULL;
    value = (value ^ (value >> 27U)) * 0x94D049BB133111EBULL;
    return value ^ (value >> 31U);
}

} // namespace

std::size_t StableOrderIdHash::operator()(OrderId order_id) const noexcept {
    return static_cast<std::size_t>(mix_order_id(order_id));
}

OrderBook::OrderBook(std::size_t expected_order_capacity) {
    orders_.reserve(expected_order_capacity);
}

void OrderBook::apply(const Event &event) {
    switch (event.type) {
    case EventType::Add:
        add_order(event.order_id, event.side, event.price_ticks, event.quantity);
        return;
    case EventType::Cancel:
        subtract_from_order(event.order_id, event.quantity, false);
        return;
    case EventType::Execute:
        subtract_from_order(event.order_id, event.quantity, true);
        return;
    case EventType::Delete:
    case EventType::OrderDelete:
        remove_order(event.order_id);
        return;
    case EventType::Replace:
        if (event.replacement_order_id == event.order_id) {
            throw std::invalid_argument("replace requires a new order identifier");
        }
        remove_order(event.order_id);
        add_order(event.replacement_order_id, event.side, event.price_ticks, event.quantity);
        return;
    case EventType::OrderUpsert:
        upsert_order(event.order_id, event.side, event.price_ticks, event.quantity);
        return;
    }
    throw std::invalid_argument("unknown event type");
}

std::optional<Level> OrderBook::best_bid() const {
    if (bids_.empty()) {
        return std::nullopt;
    }
    const auto &[price, quantity] = *bids_.rbegin();
    return Level{price, quantity};
}

std::optional<Level> OrderBook::best_ask() const {
    if (asks_.empty()) {
        return std::nullopt;
    }
    const auto &[price, quantity] = *asks_.begin();
    return Level{price, quantity};
}

std::optional<Quantity> OrderBook::level_quantity(Side side, Price price_ticks) const {
    const auto &levels = levels_for(side);
    const auto it = levels.find(price_ticks);
    return it == levels.end() ? std::nullopt : std::optional<Quantity>{it->second};
}

std::size_t OrderBook::live_order_count() const noexcept {
    return orders_.size();
}

std::size_t OrderBook::order_index_capacity() const noexcept {
    return orders_.bucket_count();
}

std::uint64_t OrderBook::state_digest() const {
    std::uint64_t hash = kFnvOffsetBasis;
    hash_u64(hash, static_cast<std::uint64_t>(orders_.size()));

    for (const auto *levels : std::array{&bids_, &asks_}) {
        hash_u64(hash, static_cast<std::uint64_t>(levels->size()));
        for (const auto &[price, quantity] : *levels) {
            hash_u64(hash, price);
            hash_u64(hash, quantity);
        }
    }

    std::vector<OrderId> order_ids;
    order_ids.reserve(orders_.size());
    for (const auto &[order_id, unused] : orders_) {
        (void)unused;
        order_ids.push_back(order_id);
    }
    std::sort(order_ids.begin(), order_ids.end());
    for (const OrderId order_id : order_ids) {
        const Order &order = orders_.at(order_id);
        hash_u64(hash, order_id);
        hash_byte(hash, static_cast<std::uint8_t>(order.side));
        hash_u64(hash, order.price_ticks);
        hash_u64(hash, order.remaining_quantity);
    }
    return hash;
}

OrderBook::Levels &OrderBook::levels_for(Side side) {
    return side == Side::Bid ? bids_ : asks_;
}

const OrderBook::Levels &OrderBook::levels_for(Side side) const {
    return side == Side::Bid ? bids_ : asks_;
}

void OrderBook::add_order(OrderId order_id, Side side, Price price_ticks, Quantity quantity) {
    if (order_id == 0 || quantity == 0 || orders_.contains(order_id)) {
        throw std::invalid_argument("invalid add order");
    }
    add_to_level(side, price_ticks, quantity);
    orders_.emplace(order_id, Order{side, price_ticks, quantity});
}

void OrderBook::remove_order(OrderId order_id) {
    const auto it = orders_.find(order_id);
    if (it == orders_.end()) {
        throw std::invalid_argument("unknown order identifier");
    }
    subtract_from_level(it->second.side, it->second.price_ticks, it->second.remaining_quantity);
    orders_.erase(it);
}

void OrderBook::subtract_from_order(OrderId order_id, Quantity quantity, bool allow_full_removal) {
    const auto it = orders_.find(order_id);
    if (it == orders_.end() || quantity == 0 || quantity > it->second.remaining_quantity ||
        (!allow_full_removal && quantity == it->second.remaining_quantity)) {
        throw std::invalid_argument("invalid order reduction");
    }
    const Order order = it->second;
    subtract_from_level(order.side, order.price_ticks, quantity);
    if (quantity == order.remaining_quantity) {
        orders_.erase(it);
    } else {
        it->second.remaining_quantity -= quantity;
    }
}

void OrderBook::upsert_order(OrderId order_id, Side side, Price price_ticks, Quantity quantity) {
    if (order_id == 0 || quantity == 0) {
        throw std::invalid_argument("invalid order upsert");
    }
    const auto it = orders_.find(order_id);
    if (it != orders_.end()) {
        subtract_from_level(it->second.side, it->second.price_ticks, it->second.remaining_quantity);
        it->second = Order{side, price_ticks, quantity};
    } else {
        orders_.emplace(order_id, Order{side, price_ticks, quantity});
    }
    add_to_level(side, price_ticks, quantity);
}

void OrderBook::add_to_level(Side side, Price price_ticks, Quantity quantity) {
    auto &level = levels_for(side)[price_ticks];
    if (UINT64_MAX - level < quantity) {
        throw std::overflow_error("price-level quantity overflow");
    }
    level += quantity;
}

void OrderBook::subtract_from_level(Side side, Price price_ticks, Quantity quantity) {
    auto &levels = levels_for(side);
    const auto it = levels.find(price_ticks);
    if (it == levels.end() || quantity == 0 || quantity > it->second) {
        throw std::logic_error("invalid price-level reduction");
    }
    it->second -= quantity;
    if (it->second == 0) {
        levels.erase(it);
    }
}

} // namespace llab::order_book

File: rust/src/order_book.rs
use std::collections::{BTreeMap, HashMap};
use std::hash::{BuildHasherDefault, Hasher};

pub type OrderId = u64;
pub type Price = u64;
pub type Quantity = u64;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Side {
    Bid,
    Ask,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Event {
    Add {
        order_id: OrderId,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    },
    Cancel {
        order_id: OrderId,
        quantity: Quantity,
    },
    Execute {
        order_id: OrderId,
        quantity: Quantity,
    },
    Delete {
        order_id: OrderId,
    },
    Replace {
        old_order_id: OrderId,
        new_order_id: OrderId,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    },
    OrderUpsert {
        order_id: OrderId,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    },
    OrderDelete {
        order_id: OrderId,
    },
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Level {
    pub price_ticks: Price,
    pub quantity: Quantity,
}

#[derive(Clone, Copy)]
struct Order {
    side: Side,
    price_ticks: Price,
    remaining_quantity: Quantity,
}

#[derive(Default)]
struct StableOrderIdHasher {
    hash: u64,
}

impl Hasher for StableOrderIdHasher {
    fn finish(&self) -> u64 {
        self.hash
    }

    fn write(&mut self, bytes: &[u8]) {
        let mut value = 14_695_981_039_346_656_037_u64;
        for &byte in bytes {
            value ^= u64::from(byte);
            value = value.wrapping_mul(1_099_511_628_211);
        }
        self.hash = mix_order_id(value);
    }

    fn write_u64(&mut self, order_id: u64) {
        self.hash = mix_order_id(order_id);
    }
}

type OrderIndex = HashMap<OrderId, Order, BuildHasherDefault<StableOrderIdHasher>>;

#[derive(Default)]
pub struct OrderBook {
    orders: OrderIndex,
    bids: BTreeMap<Price, Quantity>,
    asks: BTreeMap<Price, Quantity>,
}

impl OrderBook {
    pub fn with_order_capacity(expected_order_capacity: usize) -> Self {
        Self {
            orders: HashMap::with_capacity_and_hasher(
                expected_order_capacity,
                BuildHasherDefault::default(),
            ),
            ..Self::default()
        }
    }

    pub fn apply(&mut self, event: Event) -> Result<(), &'static str> {
        match event {
            Event::Add { order_id, side, price_ticks, quantity } => {
                self.add_order(order_id, side, price_ticks, quantity)
            }
            Event::Cancel { order_id, quantity } => {
                self.subtract_from_order(order_id, quantity, false)
            }
            Event::Execute { order_id, quantity } => {
                self.subtract_from_order(order_id, quantity, true)
            }
            Event::Delete { order_id } | Event::OrderDelete { order_id } => {
                self.remove_order(order_id)
            }
            Event::Replace { old_order_id, new_order_id, side, price_ticks, quantity } => {
                if old_order_id == new_order_id {
                    return Err("replace requires a new order identifier");
                }
                self.remove_order(old_order_id)?;
                self.add_order(new_order_id, side, price_ticks, quantity)
            }
            Event::OrderUpsert { order_id, side, price_ticks, quantity } => {
                self.upsert_order(order_id, side, price_ticks, quantity)
            }
        }
    }

    pub fn best_bid(&self) -> Option<Level> {
        self.bids.last_key_value().map(|(&price_ticks, &quantity)| Level { price_ticks, quantity })
    }

    pub fn best_ask(&self) -> Option<Level> {
        self.asks.first_key_value().map(|(&price_ticks, &quantity)| Level { price_ticks, quantity })
    }

    pub fn level_quantity(&self, side: Side, price_ticks: Price) -> Option<Quantity> {
        self.levels(side).get(&price_ticks).copied()
    }

    pub fn live_order_count(&self) -> usize {
        self.orders.len()
    }

    pub fn order_index_capacity(&self) -> usize {
        self.orders.capacity()
    }

    pub fn state_digest(&self) -> u64 {
        let mut hash = 14_695_981_039_346_656_037_u64;
        hash_u64(&mut hash, self.orders.len() as u64);
        for levels in [&self.bids, &self.asks] {
            hash_u64(&mut hash, levels.len() as u64);
            for (&price, &quantity) in levels {
                hash_u64(&mut hash, price);
                hash_u64(&mut hash, quantity);
            }
        }
        let mut order_ids: Vec<_> = self.orders.keys().copied().collect();
        order_ids.sort_unstable();
        for order_id in order_ids {
            let order = self.orders[&order_id];
            hash_u64(&mut hash, order_id);
            hash_byte(&mut hash, side_byte(order.side));
            hash_u64(&mut hash, order.price_ticks);
            hash_u64(&mut hash, order.remaining_quantity);
        }
        hash
    }

    fn add_order(
        &mut self,
        order_id: OrderId,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    ) -> Result<(), &'static str> {
        if order_id == 0 || quantity == 0 || self.orders.contains_key(&order_id) {
            return Err("invalid add order");
        }
        self.add_to_level(side, price_ticks, quantity)?;
        self.orders.insert(order_id, Order { side, price_ticks, remaining_quantity: quantity });
        Ok(())
    }

    fn remove_order(&mut self, order_id: OrderId) -> Result<(), &'static str> {
        let order = self.orders.remove(&order_id).ok_or("unknown order identifier")?;
        self.subtract_from_level(order.side, order.price_ticks, order.remaining_quantity)
    }

    fn subtract_from_order(
        &mut self,
        order_id: OrderId,
        quantity: Quantity,
        allow_full_removal: bool,
    ) -> Result<(), &'static str> {
        let order = *self.orders.get(&order_id).ok_or("unknown order identifier")?;
        if quantity == 0
            || quantity > order.remaining_quantity
            || (!allow_full_removal && quantity == order.remaining_quantity)
        {
            return Err("invalid order reduction");
        }
        self.subtract_from_level(order.side, order.price_ticks, quantity)?;
        if quantity == order.remaining_quantity {
            self.orders.remove(&order_id);
        } else if let Some(existing) = self.orders.get_mut(&order_id) {
            existing.remaining_quantity -= quantity;
        }
        Ok(())
    }

    fn upsert_order(
        &mut self,
        order_id: OrderId,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    ) -> Result<(), &'static str> {
        if order_id == 0 || quantity == 0 {
            return Err("invalid order upsert");
        }
        if let Some(existing) = self.orders.get(&order_id).copied() {
            self.subtract_from_level(
                existing.side,
                existing.price_ticks,
                existing.remaining_quantity,
            )?;
        }
        self.add_to_level(side, price_ticks, quantity)?;
        self.orders.insert(order_id, Order { side, price_ticks, remaining_quantity: quantity });
        Ok(())
    }

    fn levels(&self, side: Side) -> &BTreeMap<Price, Quantity> {
        match side {
            Side::Bid => &self.bids,
            Side::Ask => &self.asks,
        }
    }

    fn levels_mut(&mut self, side: Side) -> &mut BTreeMap<Price, Quantity> {
        match side {
            Side::Bid => &mut self.bids,
            Side::Ask => &mut self.asks,
        }
    }

    fn add_to_level(
        &mut self,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    ) -> Result<(), &'static str> {
        let level = self.levels_mut(side).entry(price_ticks).or_default();
        *level = level.checked_add(quantity).ok_or("price-level quantity overflow")?;
        Ok(())
    }

    fn subtract_from_level(
        &mut self,
        side: Side,
        price_ticks: Price,
        quantity: Quantity,
    ) -> Result<(), &'static str> {
        let levels = self.levels_mut(side);
        let level = levels.get_mut(&price_ticks).ok_or("invalid price-level reduction")?;
        if quantity == 0 || quantity > *level {
            return Err("invalid price-level reduction");
        }
        *level -= quantity;
        if *level == 0 {
            levels.remove(&price_ticks);
        }
        Ok(())
    }
}

fn side_byte(side: Side) -> u8 {
    match side {
        Side::Bid => 0,
        Side::Ask => 1,
    }
}

fn mix_order_id(mut value: u64) -> u64 {
    value = value.wrapping_add(0x9E37_79B9_7F4A_7C15);
    value = (value ^ (value >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
    value = (value ^ (value >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
    value ^ (value >> 31)
}

fn hash_byte(hash: &mut u64, byte: u8) {
    *hash ^= u64::from(byte);
    *hash = hash.wrapping_mul(1_099_511_628_211);
}

fn hash_u64(hash: &mut u64, value: u64) {
    for shift in (0..64).step_by(8) {
        hash_byte(hash, (value >> shift) as u8);
    }
}
